test/core/end2end/fixtures/h2_tls_wrong_versions.cc

```diff
@@ -0,0 +1,170 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <string.h>
+
+#include <string>
+
+#include <grpc/grpc.h>
+
+#include "src/core/lib/gprpp/global_config_generic.h"
+#include "src/core/lib/gprpp/host_port.h"
+#include "src/core/lib/security/security_connector/ssl_utils_config.h"
+#include "test/core/end2end/end2end_tests.h"
+#include "test/core/end2end/fixtures/h2_tls_common.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+
+static grpc_end2end_test_fixture chttp2_create_fixture_simple_fullstack_1(
+    const grpc_channel_args*, const grpc_channel_args*) {
+  grpc_end2end_test_fixture f;
+  int port = grpc_pick_unused_port_or_die();
+  fullstack_secure_fixture_data* ffd = new fullstack_secure_fixture_data();
+  memset(&f, 0, sizeof(f));
+  ffd->localaddr = grpc_core::JoinHostPort("localhost", port);
+  SetChannelTlsVersion(ffd, SecurityPrimitives::TlsVersion::V_12,
+                       SecurityPrimitives::TlsVersion::V_12);
+  SetServerTlsVersion(ffd, SecurityPrimitives::TlsVersion::V_13,
+                      SecurityPrimitives::TlsVersion::V_13);
+  SetCertificateProvider(ffd,
+                         SecurityPrimitives::ProviderType::STATIC_PROVIDER);
+  SetCertificateVerifier(
+      ffd, SecurityPrimitives::VerifierType::EXTERNAL_SYNC_VERIFIER);
+  f.fixture_data = ffd;
+  f.cq = grpc_completion_queue_create_for_next(nullptr);
+  return f;
+}
+
+static grpc_end2end_test_fixture chttp2_create_fixture_simple_fullstack_2(
+    const grpc_channel_args*, const grpc_channel_args*) {
+  grpc_end2end_test_fixture f;
+  int port = grpc_pick_unused_port_or_die();
+  fullstack_secure_fixture_data* ffd = new fullstack_secure_fixture_data();
+  memset(&f, 0, sizeof(f));
+  ffd->localaddr = grpc_core::JoinHostPort("localhost", port);
+  SetChannelTlsVersion(ffd, SecurityPrimitives::TlsVersion::V_13,
+                       SecurityPrimitives::TlsVersion::V_13);
+  SetServerTlsVersion(ffd, SecurityPrimitives::TlsVersion::V_12,
+                      SecurityPrimitives::TlsVersion::V_12);
+  SetCertificateProvider(ffd,
+                         SecurityPrimitives::ProviderType::STATIC_PROVIDER);
+  SetCertificateVerifier(
+      ffd, SecurityPrimitives::VerifierType::EXTERNAL_SYNC_VERIFIER);
+  f.fixture_data = ffd;
+  f.cq = grpc_completion_queue_create_for_next(nullptr);
+  return f;
+}
+
+static grpc_end2end_test_fixture chttp2_create_fixture_simple_fullstack_3(
+    const grpc_channel_args*, const grpc_channel_args*) {
+  grpc_end2end_test_fixture f;
+  int port = grpc_pick_unused_port_or_die();
+  fullstack_secure_fixture_data* ffd = new fullstack_secure_fixture_data();
+  memset(&f, 0, sizeof(f));
+  ffd->localaddr = grpc_core::JoinHostPort("localhost", port);
+  SetChannelTlsVersion(ffd, SecurityPrimitives::TlsVersion::V_13,
```

Comment:
Would it make sense to add minVersion <= maxVersion check on gRPC level? For example in src/core/lib/security/credentials/tls/grpc_tls_credentials_options.cc
`If max lvl already set, ASSERT   min_tls_version <= options->get max_tls_version
options->set_min_tls_version(min_tls_version);
`
My concern is if underlying behavior of OpenSSL changes we might introduce unpredictable behavior if we don't implement such chekc

Reply:
This is a good point but I don't think it should be the setter's job to check the versions. IIUC, tls_credentials_options is not immutable once created and can be consumed multiple times if needed. Consider the following use case:
```
options->set_min_tls_version(1.2);
options->set_max_tls_version(1.2);

// some work using the options without losing the ownership of the pointer.

// set tls versions differently for some other use.
options->set_min_tls_version(1.3); // now this would error if we check it here.
options->set_max_tls_version(1.3);
```
This admittedly is not a standard use pattern, since passing the options to the tls_credentials normally transfers the ownership of the options, but rejecting it seems weird to me.

In fact, tls_credentials_options has no functions like verify or validate. It would be better for the consumer to perform such a check, i.e., a check can be done when the options are passed into tls_credentials, either channel or server. In this case, I think it's better to be added in a separate PR.

Reply:
The snippet you described can be implemented like
`
options->set_min_tls_version(1.2);
options->set_max_tls_version(1.2);

// some work using the options without losing the ownership of the pointer.

// set tls versions differently for some other use.
options->set_min_tls_version(1.3); // Error here says 'Currently your min version is > your max version that can lead to unstable behavior of underlying SSL lib
options->set_max_tls_version(1.3);

//User changes it to

options->set_max_tls_version(1.3);
options->set_min_tls_version(1.3); 
`
But I agree that it's a better idea to perform such check at channel or server creation level.

Reply:
Thanks. In this case, I have removed those 2 fixtures where min/max versions are reversed, given that a subsequent task would be to add checks and require different tests.

For some reason, the timeout issue has disappeared when I test locally today. The test always finishes in 1.1 seconds. Are you able to reproduce it?

Reply:
I had an offline conversation with Mark to discuss the best practices for this repo. He suggested to do the check (and inform user in case of failed check) as early as possible. So I'm suggesting smth like 
`options->set_min_tls_version(1.2);
options->set_max_tls_version(1.2);

// some work using the options without losing the ownership of the pointer.

// set tls versions differently for some other use.
options->set_min_tls_version(1.3); // Error here says 'Currently your min version is > your max version that can lead to unstable behavior of underlying SSL lib
options->set_max_tls_version(1.3);

//User changes it to

options->set_max_tls_version(1.3);
options->set_min_tls_version(1.3);` 
@markdroth

Reply:
After some discussion with @erm-g, we all agree that a check is needed as early as possible. Instead of doing this at the setter, I found there already exists a [sanity check](https://github.com/grpc/grpc/blob/f76948b12d9e853cea19bbafaf098533f6907d93/src/core/lib/security/credentials/tls/tls_credentials.cc#L43) invoked right before the options are used to created the credentials. So I added a check there and some unit tests.

Reply:
There is some memory safety issue I need to handle. Will get back here when the PR is ready for a look again.

Reply:
All relevant tests should have passed. PTAL when you have time, @erm-g